Support list for Zookeeper URL for Solr cloud and index job

Reported on [janusgraph-users](https://groups.google.com/d/msg/janusgraph-users/x5CnYVpr1ls/BNfwNQ9dCAAJ)

```
15:44:17 WARN  org.apache.hadoop.hbase.zookeeper.RecoverableZooKeeper  - Unable to create ZooKeeper Connection
java.net.UnknownHostException: [app01: invalid IPv6 address
```

Where the graph configuration properties

```
index.search.solr.zookeeper-url=app01:2181,app02:2181,app03:2181
```

The [`ZOOKEEPER_URL`](https://github.com/JanusGraph/janusgraph/blob/0.2/janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java#L161-L163) is defined as a `String` property rather than a `String[]` (for example, `storage.hostname` and `index.[X].solr.http-urls` use `String[]`). If the property allows a list, it would allow for a quorum of Zookeeper nodes to be specified.

I have to change storage.hostname to only one host  to not get the error when I execute MapReduceIndexManagement. But I get different error which is:
Exception in thread "main" java.lang.NoClassDefFoundError: com/sun/jersey/api/client/config/ClientConfig
	at org.apache.hadoop.yarn.client.api.TimelineClient.createTimelineClient(TimelineClient.java:59)
	at org.apache.hadoop.yarn.client.api.impl.YarnClientImpl.createTimelineClient(YarnClientImpl.java:185)
	at org.apache.hadoop.yarn.client.api.impl.YarnClientImpl.getTimelineDelegationToken(YarnClientImpl.java:353)
	at org.apache.hadoop.yarn.client.api.impl.YarnClientImpl.addTimelineDelegationToken(YarnClientImpl.java:330)
	at org.apache.hadoop.yarn.client.api.impl.YarnClientImpl.submitApplication(YarnClientImpl.java:250)
	at org.apache.hadoop.mapred.ResourceMgrDelegate.submitApplication(ResourceMgrDelegate.java:291)
	at org.apache.hadoop.mapred.YARNRunner.submitJob(YARNRunner.java:291)
	at org.apache.hadoop.mapreduce.JobSubmitter.submitJobInternal(JobSubmitter.java:240)
	at org.apache.hadoop.mapreduce.Job$10.run(Job.java:1290)
	at org.apache.hadoop.mapreduce.Job$10.run(Job.java:1287)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1866)
	at org.apache.hadoop.mapreduce.Job.submit(Job.java:1287)
	at org.apache.hadoop.mapreduce.Job.waitForCompletion(Job.java:1308)
	at org.janusgraph.hadoop.scan.HadoopScanRunner.runJob(HadoopScanRunner.java:137)
	at org.janusgraph.hadoop.MapReduceIndexManagement.updateIndex(MapReduceIndexManagement.java:187)
	at com.att.hypergraph.core.GraphManager.reIndex(GraphManager.java:266)
	at com.att.hypergraph.core.TestGraphManagmentDriver.main(TestGraphManagmentDriver.java:41)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:750)
	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:187)
	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:212)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:126)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
Caused by: java.lang.ClassNotFoundException: com.sun.jersey.api.client.config.ClientConfig
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:338)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	... 28 more

--- a/janusgraph-hadoop-parent/janusgraph-hadoop-core/src/main/java/org/janusgraph/hadoop/MapReduceIndexManagement.java
+++ b/janusgraph-hadoop-parent/janusgraph-hadoop-core/src/main/java/org/janusgraph/hadoop/MapReduceIndexManagement.java
@@ -54,6 +54,7 @@ import org.apache.tinkerpop.gremlin.structure.Graph;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.Collection;
 import java.util.EnumSet;
 import java.util.Iterator;
 import java.util.Set;
@@ -177,7 +178,7 @@ public class MapReduceIndexManagement {
         janusgraphmrConf.set(JanusGraphHadoopConfiguration.SCAN_JOB_CONFIG_ROOT,
                 GraphDatabaseConfiguration.class.getName() + "#JOB_NS");
         // Copy the StandardJanusGraph configuration under JanusGraphHadoopConfiguration.GRAPH_CONFIG_KEYS
-        org.apache.commons.configuration.Configuration localbc = graph.getConfiguration().getLocalConfiguration();
+        org.apache.commons.configuration.Configuration localbc = graph.getConfiguration().getConfigurationAtOpen();
         localbc.clearProperty(Graph.GRAPH);
         copyInputKeys(hadoopConf, localbc);
 
@@ -195,19 +196,15 @@ public class MapReduceIndexManagement {
         Iterator<String> keyIter = source.getKeys();
         while (keyIter.hasNext()) {
             String key = keyIter.next();
-            ConfigElement.PathIdentifier pid;
-            try {
-                pid = ConfigElement.parse(ROOT_NS, key);
-            } catch (RuntimeException e) {
-                log.debug("[inputkeys] Skipping {}", key, e);
-                continue;
-            }
-
-            if (!pid.element.isOption())
-                continue;
 
             String k = ConfigElement.getPath(JanusGraphHadoopConfiguration.GRAPH_CONFIG_KEYS, true) + "." + key;
-            String v = source.getProperty(key).toString();
+            Object vObject = source.getProperty(key);
+            String v;
+            if (vObject instanceof Collection) {
+                v = Joiner.on(",").join((Collection<String>) vObject);
+            } else {
+                v = vObject.toString();
+            }
 
             hadoopConf.set(k, v);
             log.debug("[inputkeys] Set {}={}", k, v);

--- a/janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java
+++ b/janusgraph-solr/src/main/java/org/janusgraph/diskstorage/solr/SolrIndex.java
@@ -26,6 +26,7 @@ import java.text.SimpleDateFormat;
 import java.time.Instant;
 import java.util.AbstractMap.SimpleEntry;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.Date;
@@ -157,10 +158,13 @@ public class SolrIndex implements IndexProvider {
             ConfigOption.Type.GLOBAL_OFFLINE, "ttl");
 
     /** SolrCloud Configuration */
-
-    public static final ConfigOption<String> ZOOKEEPER_URL = new ConfigOption<String>(SOLR_NS,"zookeeper-url",
+    /*
+     * TODO Rename ZOOKEEPER_URL and "zookeeper-url" to ZOOKEEPER_URLS and
+     * "zookeeper-urls" in future major releases.
+     */
+    public static final ConfigOption<String[]> ZOOKEEPER_URL = new ConfigOption<String[]>(SOLR_NS,"zookeeper-url",
             "URL of the Zookeeper instance coordinating the SolrCloud cluster",
-            ConfigOption.Type.MASKABLE, "localhost:2181");
+            ConfigOption.Type.MASKABLE, new String[] { "localhost:2181" });
 
     public static final ConfigOption<Integer> NUM_SHARDS = new ConfigOption<Integer>(SOLR_NS,"num-shards",
             "Number of shards for a collection. This applies when creating a new collection which is only supported under the SolrCloud operation mode.",
@@ -239,11 +243,28 @@ public class SolrIndex implements IndexProvider {
         waitSearcher = config.get(WAIT_SEARCHER);
 
         if (mode==Mode.CLOUD) {
-            final String zookeeperUrl = config.get(SolrIndex.ZOOKEEPER_URL);
-            final CloudSolrClient cloudServer = new CloudSolrClient.Builder()
-                .withZkHost(zookeeperUrl)
-                .sendUpdatesOnlyToShardLeaders()
-                .build();
+            final String[] zookeeperUrl = config.get(SolrIndex.ZOOKEEPER_URL);
+            // Process possible zookeeper chroot. e.g. localhost:2181/solr
+            // chroot has to be the same assuming one Zookeeper ensemble.
+            // Parse from the last string. If found, take it as the chroot.
+            String chroot = null;
+            for (int i = zookeeperUrl.length - 1; i >= 0; i--) {
+                int chrootIndex = zookeeperUrl[i].indexOf("/");
+                if (chrootIndex != -1) {
+                    String hostAndPort = zookeeperUrl[i].substring(0, chrootIndex);
+                    if (chroot == null) {
+                        chroot = zookeeperUrl[i].substring(chrootIndex);
+                    }
+                    zookeeperUrl[i] = hostAndPort;
+                }
+            }
+            final CloudSolrClient.Builder builder = new CloudSolrClient.Builder()
+                    .withZkHost(Arrays.asList(zookeeperUrl))
+                    .sendUpdatesOnlyToShardLeaders();
+            if (chroot != null) {
+                builder.withZkChroot(chroot);
+            }
+            final CloudSolrClient cloudServer = builder.build();
             cloudServer.connect();
             solrClient = cloudServer;
         } else if (mode==Mode.HTTP) {

--- a/janusgraph-solr/src/test/java/org/janusgraph/diskstorage/solr/SolrRunner.java
+++ b/janusgraph-solr/src/test/java/org/janusgraph/diskstorage/solr/SolrRunner.java
@@ -15,6 +15,8 @@
 package org.janusgraph.diskstorage.solr;
 
 import com.google.common.base.Joiner;
+import com.google.common.base.Strings;
+
 import org.apache.commons.io.FileUtils;
 import org.apache.solr.cloud.MiniSolrCloudCluster;
 
@@ -30,7 +32,7 @@ import java.util.stream.Collectors;
 
 public class SolrRunner {
 
-    public static final String ZOOKEEPER_URLS = System.getProperty("index.search.solr.zookeeper-url");
+    public static final String ZOOKEEPER_URLS_SYSTEM_PROPERTY = System.getProperty("index.search.solr.zookeeper-url");
 
     protected static final int NUM_SERVERS = 1;
     protected static final String[] COLLECTIONS = readCollections();
@@ -44,7 +46,7 @@ public class SolrRunner {
     private static MiniSolrCloudCluster miniSolrCloudCluster;
 
     public static void start() throws Exception {
-        if (ZOOKEEPER_URLS != null) {
+        if (ZOOKEEPER_URLS_SYSTEM_PROPERTY != null) {
             return;
         }
         String userDir = System.getProperty("user.dir");
@@ -73,18 +75,18 @@ public class SolrRunner {
         }
     }
 
-    public static String getZookeeperUrls() {
-        final String zookeeperUrls;
-        if (ZOOKEEPER_URLS == null) {
-            zookeeperUrls = miniSolrCloudCluster.getZkServer().getZkAddress();
+    public static String[] getZookeeperUrls() {
+        final String[] zookeeperUrls;
+        if (Strings.isNullOrEmpty(ZOOKEEPER_URLS_SYSTEM_PROPERTY)) {
+            zookeeperUrls = new String[] { miniSolrCloudCluster.getZkServer().getZkAddress() };
         } else {
-            zookeeperUrls = ZOOKEEPER_URLS;
+            zookeeperUrls = ZOOKEEPER_URLS_SYSTEM_PROPERTY.split(",");
         }
         return zookeeperUrls;
     }
 
     public static void stop() throws Exception {
-        if (ZOOKEEPER_URLS != null) {
+        if (ZOOKEEPER_URLS_SYSTEM_PROPERTY != null) {
             return;
         }
         System.clearProperty("solr.solrxml.location");

